File: packages/socket/src/handlers/ai.ts
import { AI, EVENTS } from "@razzia/common/constants"
import {
  aiGenerateDistractorsValidator,
  aiGenerateQuestionValidator,
  aiGenerateQuizValidator,
  aiSetKeyValidator,
  aiSettingsValidator,
  aiTestValidator,
} from "@razzia/common/validators/ai"
import { getClientId } from "@razzia/socket/handlers/imageGenThrottle"
import type { SocketContext } from "@razzia/socket/handlers/types"
import {
  generateDistractors,
  generateQuestion,
  generateQuiz,
  generateText,
} from "@razzia/socket/services/ai-provider"
import { setKey } from "@razzia/socket/services/ai-secrets"
import {
  getAISettings,
  setAISettings,
  toPublicAISettings,
} from "@razzia/socket/services/config"
import manager from "@razzia/socket/services/manager"

// Text-generation throttle. Text gen can spend money via a cloud key (handlers
// are auth-gated, but a logged-in manager could still hammer it), so every
// generation event shares one cooldown + lifetime cap. Keyed by the DURABLE
// clientId (services/manager.ts#getClientId), NOT socket.id, so a client cannot
// reset its cooldown / lifetime cap by reconnecting. Entries self-expire via a
// lazy time-window sweep on every read/write (no disconnect cleanup), so the
// Map cannot grow unbounded across many distinct clients.
interface TextGenState {
  last: number
  total: number
}

const textGenStore = new Map<string, TextGenState>()

// Lazy GC: drop entries whose last activity is older than this window so the
// Map cannot leak across many distinct clients. Far longer than the cooldown so
// the lifetime cap stays durable for an active client.
const TEXT_GEN_GC_MS = 3_600_000

const sweepTextGenStore = (now: number): void => {
  for (const [key, state] of textGenStore) {
    if (now - state.last > TEXT_GEN_GC_MS) {
      textGenStore.delete(key)
    }
  }
}

// Returns true if the call is allowed (and records it); false if throttled.
const allowTextGen = (clientId: string): boolean => {
  const now = Date.now()
  sweepTextGenStore(now)
  const state = textGenStore.get(clientId)

  if (state) {
    if (now - state.last < AI.TEXT_GEN_COOLDOWN_MS) {
      return false
    }

    if (state.total >= AI.TEXT_GEN_MAX_PER_SOCKET) {
      return false
    }

    state.last = now
    state.total += 1

    return true
  }

  textGenStore.set(clientId, { last: now, total: 1 })

  return true
}

const errMessage = (error: unknown, fallback: string): string =>
  error instanceof Error ? error.message : fallback

export const aiSocketHandlers = ({ socket }: SocketContext) => {
  // Read current settings (public shape — never carries a key).
  socket.on(
    EVENTS.AI.GET_SETTINGS,
    manager.withAuth(socket, () => {
      socket.emit(EVENTS.AI.SETTINGS, toPublicAISettings(getAISettings()))
    }),
  )

  socket.on(
    EVENTS.AI.SET_SETTINGS,
    manager.withAuth(socket, (payload: unknown) => {
      const result = aiSettingsValidator.safeParse(payload)

      if (!result.success) {
        socket.emit(EVENTS.AI.ERROR, result.error.issues[0].message)

        return
      }

      try {
        setAISettings(result.data)
        socket.emit(EVENTS.AI.SET_SETTINGS_SUCCESS)
        socket.emit(EVENTS.AI.SETTINGS, toPublicAISettings(getAISettings()))
      } catch (error) {
        socket.emit(EVENTS.AI.ERROR, errMessage(error, "errors:ai.saveFailed"))
      }
    }),
  )

  socket.on(
    EVENTS.AI.SET_KEY,
    manager.withAuth(socket, (payload: unknown) => {
      const result = aiSetKeyValidator.safeParse(payload)

      if (!result.success) {
        socket.emit(EVENTS.AI.ERROR, result.error.issues[0].message)

        return
      }

      try {
        // Empty/whitespace key clears the stored secret (setKey handles it).
        setKey(result.data.providerId, result.data.key.trim() || null)
        // Re-emit public settings so the client's keyConfigured flag updates.
        socket.emit(EVENTS.AI.SETTINGS, toPublicAISettings(getAISettings()))
      } catch (error) {
        socket.emit(EVENTS.AI.ERROR, errMessage(error, "errors:ai.saveFailed"))
      }
    }),
  )

  socket.on(
    EVENTS.AI.TEST_PROVIDER,
    manager.withAuth(socket, (payload: unknown) => {
      void (async () => {
        if (!allowTextGen(getClientId(socket))) {
          socket.emit(EVENTS.AI.TEST_RESULT, {
            ok: false,
            message: "errors:ai.rateLimited",
          })

          return
        }

        const result = aiTestValidator.safeParse(payload)

        if (!result.success) {
          socket.emit(EVENTS.AI.TEST_RESULT, {
            ok: false,
            message: result.error.issues[0].message,
          })

          return
        }

        try {
          // Minimal connectivity probe against the active provider.
          await generateText({ prompt: "ping", maxTokens: 5 })
          socket.emit(EVENTS.AI.TEST_RESULT, {
            ok: true,
            message: "manager:ai.testOk",
          })
        } catch (error) {
          socket.emit(EVENTS.AI.TEST_RESULT, {
            ok: false,
            message: errMessage(error, "errors:ai.providerError"),
          })
        }
      })()
    }),
  )

  socket.on(
    EVENTS.AI.GENERATE_QUESTION,
    manager.withAuth(socket, (payload: unknown) => {
      void (async () => {
        if (!allowTextGen(getClientId(socket))) {
          socket.emit(EVENTS.AI.ERROR, "errors:ai.rateLimited")

          return
        }

        const result = aiGenerateQuestionValidator.safeParse(payload)

        if (!result.success) {
          socket.emit(EVENTS.AI.ERROR, result.error.issues[0].message)

          return
        }

        try {
          const question = await generateQuestion(
            result.data.topic,
            result.data.type ?? "choice",
            result.data.language ?? "de",
          )
          socket.emit(EVENTS.AI.QUESTION_GENERATED, { question })
        } catch (error) {
          socket.emit(
            EVENTS.AI.ERROR,
            errMessage(error, "errors:ai.invalidOutput"),
          )
        }
      })()
    }),
  )

  socket.on(
    EVENTS.AI.GENERATE_DISTRACTORS,
    manager.withAuth(socket, (payload: unknown) => {
      void (async () => {
        if (!allowTextGen(getClientId(socket))) {
          socket.emit(EVENTS.AI.ERROR, "errors:ai.rateLimited")

          return
        }

        const result = aiGenerateDistractorsValidator.safeParse(payload)

        if (!result.success) {
          socket.emit(EVENTS.AI.ERROR, result.error.issues[0].message)

          return
        }

        try {
          const distractors = await generateDistractors(
            result.data.question,
            result.data.correct,
            result.data.count ?? 3,
            result.data.language ?? "de",
          )
          socket.emit(EVENTS.AI.DISTRACTORS_GENERATED, { distractors })
        } catch (error) {
          socket.emit(
            EVENTS.AI.ERROR,
            errMessage(error, "errors:ai.invalidOutput"),
          )
        }
      })()
    }),
  )

  socket.on(
    EVENTS.AI.GENERATE_QUIZ,
    manager.withAuth(socket, (payload: unknown) => {
      void (async () => {
        if (!allowTextGen(getClientId(socket))) {
          socket.emit(EVENTS.AI.ERROR, "errors:ai.rateLimited")

          return
        }

        const result = aiGenerateQuizValidator.safeParse(payload)

        if (!result.success) {
          socket.emit(EVENTS.AI.ERROR, result.error.issues[0].message)

          return
        }

        try {
          // Do NOT persist — the web client saves via QUIZZ.SAVE.
          const quizz = await generateQuiz(
            result.data.topic,
            result.data.count,
            result.data.language ?? "de",
          )
          socket.emit(EVENTS.AI.QUIZ_GENERATED, { quizz })
        } catch (error) {
          socket.emit(
            EVENTS.AI.ERROR,
            errMessage(error, "errors:ai.invalidOutput"),
          )
        }
      })()
    }),
  )
}

File: rust/server/src/socket/ai.rs
//! AI provider configuration + text generation handlers (6 socket events).
//!
//! All handlers are auth-gated (manager-only) and text-gen handlers use a shared
//! rate limiter (per-client cooldown + lifetime cap, keyed by durable client ID).
//!
//! Integrates with ai_secrets, ai_provider, and ai_ratelimit modules.

use super::HandlerCtx;
use razzoozle_protocol::constants;
use socketioxide::extract::{Data, SocketRef};

pub fn register(socket: &SocketRef, ctx: HandlerCtx) {
    // ---- GET_SETTINGS ----
    socket.on(constants::ai::GET_SETTINGS, {
        let ctx = ctx.clone();
        move |socket: SocketRef| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                socket
                    .emit(constants::ai::SETTINGS, &super::ai_config::get_public_ai_settings())
                    .ok();
            });
        }
    });

    // ---- SET_SETTINGS ----
    socket.on(constants::ai::SET_SETTINGS, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                // Validate payload
                if let Err(err) = super::ai_validate::validate_set_settings(&payload) {
                    socket.emit(constants::ai::ERROR, &err).ok();
                    return;
                }

                match super::ai_config::persist_ai_settings(&payload).await {
                    Ok(_) => {
                        socket.emit(constants::ai::SET_SETTINGS_SUCCESS, &"").ok();
                        socket
                            .emit(constants::ai::SETTINGS, &super::ai_config::get_public_ai_settings())
                            .ok();
                    }
                    Err(e) => {
                        socket.emit(constants::ai::ERROR, &e).ok();
                    }
                }
            });
        }
    });

    // ---- SET_KEY ----
    socket.on(constants::ai::SET_KEY, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_admin().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                // Validate and extract payload
                let (provider_id, key) = match super::ai_validate::validate_set_key(&payload) {
                    Ok(result) => result,
                    Err(err) => {
                        socket.emit(constants::ai::ERROR, &err).ok();
                        return;
                    }
                };

                match super::ai_secrets::set_key(&provider_id, key) {
                    Ok(()) => {
                        socket
                            .emit(constants::ai::SETTINGS, &super::ai_config::get_public_ai_settings())
                            .ok();
                    }
                    Err(e) => {
                        socket.emit(constants::ai::ERROR, &e).ok();
                    }
                }
            });
        }
    });

    // ---- TEST_PROVIDER ----
    socket.on(constants::ai::TEST_PROVIDER, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                if !super::ai_ratelimit::allow_text_gen(
                    &ctx.client_id,
                    constants::AI::TEXT_GEN_COOLDOWN_MS,
                    constants::AI::TEXT_GEN_MAX_PER_SOCKET,
                ) {
                    socket
                        .emit(
                            constants::ai::TEST_RESULT,
                            &serde_json::json!({
                                "ok": false,
                                "message": "errors:ai.rateLimited"
                            }),
                        )
                        .ok();
                    return;
                }

                // Validate payload (emit TEST_RESULT on failure, not ERROR)
                if let Err(err) = super::ai_validate::validate_test_provider(&payload) {
                    socket
                        .emit(
                            constants::ai::TEST_RESULT,
                            &serde_json::json!({
                                "ok": false,
                                "message": err
                            }),
                        )
                        .ok();
                    return;
                }

                match super::ai_provider::generate_text(super::ai_provider::GenerateTextOptions {
                    system: None,
                    prompt: "ping".to_string(),
                    json: false,
                    max_tokens: Some(5),
                })
                .await
                {
                    Ok(_) => {
                        socket
                            .emit(
                                constants::ai::TEST_RESULT,
                                &serde_json::json!({
                                    "ok": true,
                                    "message": "manager:ai.testOk"
                                }),
                            )
                            .ok();
                    }
                    Err(e) => {
                        socket
                            .emit(
                                constants::ai::TEST_RESULT,
                                &serde_json::json!({
                                    "ok": false,
                                    "message": e
                                }),
                            )
                            .ok();
                    }
                }
            });
        }
    });

    // ---- GENERATE_QUESTION ----
    socket.on(constants::ai::GENERATE_QUESTION, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                if !super::ai_ratelimit::allow_text_gen(
                    &ctx.client_id,
                    constants::AI::TEXT_GEN_COOLDOWN_MS,
                    constants::AI::TEXT_GEN_MAX_PER_SOCKET,
                ) {
                    socket.emit(constants::ai::ERROR, &"errors:ai.rateLimited").ok();
                    return;
                }

                // Validate and extract payload
                let (topic, q_type, language) = match super::ai_validate::validate_generate_question(&payload) {
                    Ok(result) => result,
                    Err(err) => {
                        socket.emit(constants::ai::ERROR, &err).ok();
                        return;
                    }
                };

                match super::ai_provider::generate_question(&topic, &q_type, &language).await {
                    Ok(question) => {
                        socket
                            .emit(
                                constants::ai::QUESTION_GENERATED,
                                &serde_json::json!({ "question": question }),
                            )
                            .ok();
                    }
                    Err(e) => {
                        socket.emit(constants::ai::ERROR, &e).ok();
                    }
                }
            });
        }
    });

    // ---- GENERATE_DISTRACTORS ----
    socket.on(constants::ai::GENERATE_DISTRACTORS, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                if !super::ai_ratelimit::allow_text_gen(
                    &ctx.client_id,
                    constants::AI::TEXT_GEN_COOLDOWN_MS,
                    constants::AI::TEXT_GEN_MAX_PER_SOCKET,
                ) {
                    socket.emit(constants::ai::ERROR, &"errors:ai.rateLimited").ok();
                    return;
                }

                // Validate and extract payload
                let (question, correct, count, language) = match super::ai_validate::validate_generate_distractors(&payload) {
                    Ok(result) => result,
                    Err(err) => {
                        socket.emit(constants::ai::ERROR, &err).ok();
                        return;
                    }
                };

                match super::ai_provider::generate_distractors(&question, &correct, count, &language)
                    .await
                {
                    Ok(distractors) => {
                        socket
                            .emit(
                                constants::ai::DISTRACTORS_GENERATED,
                                &serde_json::json!({ "distractors": distractors }),
                            )
                            .ok();
                    }
                    Err(e) => {
                        socket.emit(constants::ai::ERROR, &e).ok();
                    }
                }
            });
        }
    });

    // ---- GENERATE_QUIZ ----
    socket.on(constants::ai::GENERATE_QUIZ, {
        let ctx = ctx.clone();
        move |socket: SocketRef, Data::<serde_json::Value>(payload)| {
            let ctx = ctx.clone();
            tokio::spawn(async move {
                let _user = match ctx.require_user().await {
                    Some(user) => user,
                    None => {
                        socket
                            .emit(constants::manager::UNAUTHORIZED, &"")
                            .ok();
                        return;
                    }
                };

                if !super::ai_ratelimit::allow_text_gen(
                    &ctx.client_id,
                    constants::AI::TEXT_GEN_COOLDOWN_MS,
                    constants::AI::TEXT_GEN_MAX_PER_SOCKET,
                ) {
                    socket.emit(constants::ai::ERROR, &"errors:ai.rateLimited").ok();
                    return;
                }

                // Validate and extract payload
                let (topic, count, language) = match super::ai_validate::validate_generate_quiz(&payload) {
                    Ok(result) => result,
                    Err(err) => {
                        socket.emit(constants::ai::ERROR, &err).ok();
                        return;
                    }
                };

                match super::ai_provider::generate_quiz(&topic, count, &language).await {
                    Ok(quizz) => {
                        socket
                            .emit(
                                constants::ai::QUIZ_GENERATED,
                                &serde_json::json!({ "quizz": quizz }),
                            )
                            .ok();
                    }
                    Err(e) => {
                        socket.emit(constants::ai::ERROR, &e).ok();
                    }
                }
            });
        }
    });
}
